### PR TITLE
OpenMapTiles 3.15.0 SNAPSHOT

### PR DESCRIPTION
**Pull request by phanecak-maptiler**
This PR brings into `planetiler-openmaptiles` all the changes from OpenMapTiles ( https://github.com/openmaptiles/openmaptiles ) since 3.14.0, e.g. after it is merged, `planetiler-openmaptiles/main` should be equivalent to (current) `openmaptiles/master`. In other words, this implements OpenMapTiles 3.15.0 "snapshot".

Depending on how it goes with OpenMapTiles, the plan would be to release `planetiler-openmaptiles 3.15.0` shortly after release of `OpenMapTiles 3.15.0`.

This PR is the result of several incremental updates, usually trying to implement one OpenMapTiles PR in "1 to 1" fashion as one `planetiler-openmaptiles` PR:

* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/3
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/4
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/5
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/6
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/7
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/8
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/9
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/10
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/11
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/12
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/13
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/15
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/16
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/17
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/18
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/19
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/20
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/22
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/23
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/29
* https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/30

### CODE REVIEW

**Comment by phanecak-maptiler** [comment]
Converting to "draft" given failure in "CI / Regenerate".

**Comment by phanecak-maptiler** [comment]
> Converting to "draft" given failure in "CI / Regenerate".

CI issue adjusted, PR "open".

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Transportation.java:674]: One way that planetiler-openmaptiles does differ from openmaptiles is that pt-omt includes more detailed geometries by default, but you can change this with the `--simplify-tolerance` argument. I noticed when looking at 3.15 vs 3.14 output that roads look "blockier" in 3.15. Could we make this simplify tolerance use the default values from arguments like `Boundary` layer does?

https://github.com/openmaptiles/planetiler-openmaptiles/blob/827673b4ea20eb1a802c34eed4580ef7fe5f8bc3/src/main/java/org/openmaptiles/layers/Boundary.java#L354-L359

**Comment by msbarry** [comment]
For future reference, comparing diffs against: https://github.com/openmaptiles/openmaptiles/compare/v3.14..master#diff-eedcd88ff95a9b072a122100809c2548385cad6635b2f114ecf4af51a1c33f80

File: src/main/java/org/openmaptiles/layers/Housenumber.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Housenumber.java:85]: Minor: would be a little easier to follow. OSM_SEPARATOR would need to be a string

```suggestion
    if (!housenumber.contains(OSM_SEPARATOR)) {
```

File: src/main/java/org/openmaptiles/layers/Housenumber.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Housenumber.java:90]: Do we want to trim the output?

```suggestion
      .map(String::trim)
      .filter(n -> !n.isEmpty())
```

File: src/main/java/org/openmaptiles/layers/Housenumber.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Housenumber.java:89]: Since we're using java 21 now, if you make this a list with `stream.toList()` then later on when you need the first/last values you can use the new methods `.getFirst()` and `.getLast()`

File: src/main/java/org/openmaptiles/layers/Housenumber.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Housenumber.java:114]: I see that openmaptiles [uses street and block_number tags now](https://github.com/openmaptiles/openmaptiles/compare/v3.14..master#diff-281b2691711d0047d6db7b21fe9284dbf748a7669f91411465be94f76c43cbfdR23) although I can't quite decipher why? Do we need to do anything with them here?

**Comment by msbarry** [comment]
Do we need any change to the water layer to reflect this change:

https://github.com/openmaptiles/openmaptiles/compare/v3.14..master#diff-fe81bf218fb5ab71b7b484638d5de80e72c920bbe7dff56a336228aaa2b6b8c5R4

Or is it picked up automatically by the mapping yaml?

File: src/main/java/org/openmaptiles/util/Utils.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/util/Utils.java:117]: You can use `Math.clamp(value, min, max)` now that we've upgraded to java 21

File: src/main/java/org/openmaptiles/layers/WaterName.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/WaterName.java:214]: TODO for planetiler: "include a point at the the center of this polygon when the polygon's tile area is at least X pixels" seems like a common use-case that should be pushed down into planetiler.  Maybe reuse `.setMinPixelSize(...)` which currently doesn't have any meaning for points or a new feature method: `.whenFeatureSizeExceeds(...)`

File: src/main/java/org/openmaptiles/layers/WaterName.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/WaterName.java:211]: Based on this logic:

https://github.com/openmaptiles/openmaptiles/compare/v3.14..master#diff-201daa1c61c99073fe3280d440c9feca5ed2236b251ad454caa14cc203f952d1R54-R56

Don't we want minzoom to be set by area for sea or ocean as well?

File: src/main/java/org/openmaptiles/layers/WaterName.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/WaterName.java:204]: Based on this logic, do we always want to set minZoom = 3 when lake centerline is present?

```diff
FROM osm_water_lakeline
  WHERE geometry && bbox
-   AND ((zoom_level BETWEEN 9 AND 13 AND LineLabel(zoom_level, NULLIF(name, ''), geometry))
+   AND ((zoom_level BETWEEN 3 AND 13 AND LineLabel(zoom_level, NULLIF(name, ''), geometry))
      OR (zoom_level >= 14))
  UNION ALL
```

https://github.com/openmaptiles/openmaptiles/compare/v3.14..master#diff-201daa1c61c99073fe3280d440c9feca5ed2236b251ad454caa14cc203f952d1R32-R36

**Review by msbarry — COMMENTED** [review]
Reviewed water_name

File: src/main/java/org/openmaptiles/layers/Poi.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Poi.java:220]: Now that `agg_stop` is implemented, can we remove the line in the README that says it's a discrepancy between planetiler-openmaptiles and openmaptiles?

**Review by msbarry — COMMENTED** [review]
Reviewed POI layer and agg_stop implementation

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Transportation.java:414]: Could we add these strings like `"omt-ie-motorway"` as an optional argument to RouteNetwork enum, so we could just do:

```java
Transportation.RouteNetwork networkType = switch(highway) {
  case "motorway" -> RouteNetwork.IE_MOTORWAY;
  case "trunk", "primary" ->  RouteNetwork.IE_NATIONAL;
  default -> RouteNetwork.IE_REGIONAL;
};
result.add(new RouteRelation(refMatcher.group(), networkType.network, networkType, (byte) -1, 0));
```

And same or GB?

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Transportation.java:526]: could `TRUNK_AS_MOTORWAY_BY_NETWORK` store the enum values and compare against them instead of the strings?

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Transportation.java:541]: Does this mean the min zoom that brunnel appears at is 9? Looking at the [OMT diff](https://github.com/openmaptiles/openmaptiles/compare/v3.14..master) and this PR https://github.com/openmaptiles/openmaptiles/pull/1283 it looks like brunnel should appear down to z4 but only on segments with a min length.

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Transportation.java:473]: I can't find where in the [OMT diff](https://github.com/openmaptiles/openmaptiles/compare/v3.14..master) we want to only show layer if brunnel is present? From what I can tell from the diff and https://github.com/openmaptiles/openmaptiles/pull/1283 we want brunnel to show up if >3px down to z4, and layer to show up if >3px down to z9

**Review by msbarry — COMMENTED** [review]
Reviewed transportation layer

**Review by msbarry — COMMENTED** [review]
Tested running over planet, [output here](https://protomaps.github.io/PMTiles/?url=https%3A%2F%2Fprotomaps.dev%2F~msbarry%2Fomt-3.15.pmtiles)

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [reply src/main/java/org/openmaptiles/layers/Transportation.java:541]: cc/ @ZeLonewolf who made the change in openmaptiles

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [reply src/main/java/org/openmaptiles/layers/Transportation.java:473]: cc/ @ZeLonewolf who made the change in openmaptiles

**Comment by msbarry** [comment]
OK finished reviewing! Is the intention that we wait until openmaptiles 3.15 is released to land this? Or land it first, then update `scripts/regenerate-openmaptiles.sh` back to point at 3.15 after openmaptiles 3.15 release?

File: src/main/java/org/openmaptiles/layers/WaterName.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/WaterName.java]: Yes, we do. I've botched following: https://github.com/openmaptiles/openmaptiles/pull/1457/files#diff-201daa1c61c99073fe3280d440c9feca5ed2236b251ad454caa14cc203f952d1R55

Thanks, adjusted.

File: src/main/java/org/openmaptiles/util/Utils.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/util/Utils.java]: Adjusted.

File: src/test/java/org/openmaptiles/layers/PoiTest.java
Comment (phanecak-maptiler) [reply src/test/java/org/openmaptiles/layers/PoiTest.java]: Adjusted.

**Comment by phanecak-maptiler** [comment]
> OK finished reviewing! Is the intention that we wait until openmaptiles 3.15 is released to land this? Or land it first, then update `scripts/regenerate-openmaptiles.sh` back to point at 3.15 after openmaptiles 3.15 release?

I believe we can merge right away, once review feedback is resolved. Then we will have master = main parity between OpenMapTiles and planetiler-openmaptiles.

At that point it would be nice to either release right away both or agree what from the backlog still to implement/merge and release once that is done.

So, I'll get back to the review feedback (and a conflict currently flagged for this PR). All others may either help here or check the backlog in OpenMapTiles or planetiler-openmaptiles.

File: src/main/java/org/openmaptiles/layers/WaterName.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/WaterName.java]: back-link: https://github.com/onthegomap/planetiler/issues/716

This looks nicer, hence I'll adjust the PR using the new core feature.

File: src/main/java/org/openmaptiles/layers/Housenumber.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/Housenumber.java:121]: Porting done here: https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/23

Already merged also into this.

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Transportation.java:445]: FYI @phanecak-maptiler I implemented setAttrWithMinzoom in https://github.com/onthegomap/planetiler/pull/725 - you should be able to do this now:

```java
.setAttrWithMinSize(Fields.BRUNNEL, brunnelValue, 3, 4, 12) // min size=3px starting at z4, then always show at z12 and above
```

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [reply src/main/java/org/openmaptiles/layers/Transportation.java:609]: setMinzoom+setMinPixelSizeBellowZoom on the feature should limit feature size before merging. What output were you seeing that made it look like it was limiting post-merge?

File: src/main/java/org/openmaptiles/layers/TransportationName.java
Comment (msbarry) [reply src/main/java/org/openmaptiles/layers/TransportationName.java:282]: Implemented in https://github.com/onthegomap/planetiler/pull/725

File: src/main/java/org/openmaptiles/layers/WaterName.java
Comment (msbarry) [reply src/main/java/org/openmaptiles/layers/WaterName.java:204]: OK this might end up behaving exactly the same way as setting minzoom=3 if centerline is present because the class should always "lake."  So as long as there aren't any centerlines in the dataset for "bay" or "sea" then it should be the same.

File: src/main/java/org/openmaptiles/layers/Poi.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/Poi.java:230]: Checked and yes, there were some stops lumped together even if far apart. Most extreme case: https://www.openstreetmap.org/node/1861342961 vs. https://www.openstreetmap.org/node/2037050203 + https://www.openstreetmap.org/node/2037050384

Adjusted with https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/24 and https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/25 .

File: src/main/java/org/openmaptiles/layers/Poi.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/Poi.java:246]: Checked and yes, there were some stops lumped together even if far apart. Most extreme case: https://www.openstreetmap.org/node/1861342961 vs. https://www.openstreetmap.org/node/2037050203 + https://www.openstreetmap.org/node/2037050384

Adjusted with https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/24 and https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/25 .

File: src/main/java/org/openmaptiles/layers/Poi.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/Poi.java:231]: Adjusted with https://github.com/phanecak-maptiler/planetiler-openmaptiles/pull/26 , merged here.

File: src/main/java/org/openmaptiles/layers/TransportationName.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/TransportationName.java:281]: Default is still there, e.g. omit brunnel tag which deviates from what OMT dies. Hence README can IMHO stay as is.

What is inside the `if` should be closer to what OMT does, if running with the default changed.

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [reply src/main/java/org/openmaptiles/layers/Transportation.java]: Did you try this out?

```java
  @Override
  public void process(Tables.OsmShipwayLinestring element, FeatureCollector features) {
    features.line(LAYER_NAME).setBufferPixels(BUFFER_SIZE)
      .setAttr(Fields.CLASS, element.shipway()) // "ferry"
      // no subclass
      .setAttr(Fields.SERVICE, service(element.service()))
      .setAttr(Fields.RAMP, element.isRamp() ? 1L : null)
      .setAttr(Fields.BRUNNEL, brunnel(element.isBridge(), element.isTunnel(), element.isFord()))
      .setAttr(Fields.LAYER, nullIfLong(element.layer(), 0))
      .setSortKey(element.zOrder())
      .setMinPixelSize(0) // merge during post-processing, then limit by size
      .setMinZoom(4)
      .setMinPixelSizeBelowZoom(10, 32);
  }
```

that should be doing exactly the same thing as the `getFerryMinzoom` function.  When I try that out I see:

<table>
<tr>
<td>
Your branch:
image

</td>
<td>
with that change:
image

</td>
</tr>
</table>

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [inline src/main/java/org/openmaptiles/layers/Transportation.java:602]: re-linking more discussion here: https://github.com/openmaptiles/planetiler-openmaptiles/pull/126#discussion_r1390392894

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (msbarry) [reply src/main/java/org/openmaptiles/layers/Transportation.java]: Thanks!! It might be worth playing with the planetiler settings a bit to see what value the default tolerance gets as close as possible to openmaptiles - we could even make that the default, but it's good for whoever's running it to be able to configure.

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/Transportation.java:277]: adjusted based on first suggestion

File: src/main/java/org/openmaptiles/layers/Poi.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/Poi.java]: Adjusted.

Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/Poi.java]: Adjusted. (Not sure why I see your comment twice, hence also answering twice to make sure it won't block review.)

File: src/main/java/org/openmaptiles/layers/Transportation.java
Comment (phanecak-maptiler) [reply src/main/java/org/openmaptiles/layers/Transportation.java:588]: Adjusted.

File: src/test/java/org/openmaptiles/layers/TransportationTest.java
Comment (msbarry) [inline src/test/java/org/openmaptiles/layers/TransportationTest.java:1813]: This test seems to be failing, I think we just need to change to this to capture the minPixelSize change:

```suggestion
      "_minzoom", 4,
      "_maxzoom", 14,
      "_minpixelsize", 32d,
```

**Review by msbarry — APPROVED** [review]

**Comment by msbarry** [comment]
Once that test is fixed, looks good to merge from my end. Thanks for the hard work on this!

File: src/test/java/org/openmaptiles/layers/TransportationTest.java
Comment (phanecak-maptiler) [reply src/test/java/org/openmaptiles/layers/TransportationTest.java]: Thank you, adjusted.

**Pull request merged**
